### PR TITLE
[lldb][NativePDB] Parse global variables.

### PR DESCRIPTION
**Pull request by ZequanWu**
This doesn't parse S_CONSTANT case yet, because I found that the global variable `std::strong_ordering::equal` is a S_CONSTANT and has type of LF_STRUCTURE which is not currently handled when creating dwarf expression for the variable. Left a TODO for it to finish later.

This makes `lldb/test/Shell/SymbolFile/PDB/ast-restore.test` and  `lldb/test/Shell/SymbolFile/PDB/calling-conventions-x86.test` pass on windows with native pdb plugin only.

### CODE REVIEW

File: lldb/source/Plugins/SymbolFile/NativePDB/SymbolFileNativePDB.cpp [hunk 1]
```diff
@@ -365,18 +365,20 @@ void SymbolFileNativePDB::InitializeObject() {
 }
 
 uint32_t SymbolFileNativePDB::CalculateNumCompileUnits() {
+  if (m_cu_count)
+    return *m_cu_count;
```
Comment (labath) [inline lldb/source/Plugins/SymbolFile/NativePDB/SymbolFileNativePDB.cpp:369]: `CalculateNumCompileUnits` is wrapped by `GetNumCompileUnits`, which makes sure the computation happens only once. I think you can delete this variable and just call `GetNumCompileUnits` everywhere.

File: lldb/source/Plugins/SymbolFile/NativePDB/SymbolFileNativePDB.cpp [hunk 2]
```diff
@@ -1810,7 +1813,27 @@ SymbolFileNativePDB::ParseVariablesForCompileUnit(CompileUnit &comp_unit,
...
   PdbSymUid sym_uid(comp_unit.GetID());
   lldbassert(sym_uid.kind() == PdbSymUidKind::Compiland);
-  return 0;
+  for (const uint32_t gid : m_index->globals().getGlobalsTable()) {
+    PdbGlobalSymId global{gid, false};
+    CVSymbol sym = m_index->ReadSymbolRecord(global);
+    // TODO: Handle S_CONSTANT which might be a record type (e.g.
+    // std::strong_ordering::equal). Currently
+    // lldb_private::npdb::MakeConstantLocationExpression doesn't handle this
+    // case and will crash if we do create global variables from it.
```
Comment (labath) [inline lldb/source/Plugins/SymbolFile/NativePDB/SymbolFileNativePDB.cpp:1822]: Is there some simple fix which would make it not crash at least?

Comment (ZequanWu) [reply lldb/source/Plugins/SymbolFile/NativePDB/SymbolFileNativePDB.cpp:1820]: Just what I did: not handling the S_CONSTANT case prevents the call to `lldb_private::npdb::MakeConstantLocationExpression` with a record type, hence not crash for now.

**Review by rnk — APPROVED** [review]

File: lldb/source/Plugins/SymbolFile/NativePDB/SymbolFileNativePDB.cpp [hunk 3]
```diff
@@ -888,7 +888,8 @@ VariableSP SymbolFileNativePDB::CreateGlobalVariable(PdbGlobalSymId var_id) {
 
   CompUnitSP comp_unit;
   std::optional<uint16_t> modi = m_index->GetModuleIndexForVa(addr);
-  if (!modi) {
+  // Some globals has modi points to the linker module, ignore them.
+  if (!modi || modi >= GetNumCompileUnits()) {
     return nullptr;
   }
```
Comment (JDevlieghere) [inline lldb/source/Plugins/SymbolFile/NativePDB/SymbolFileNativePDB.cpp:894]: Nit: no braces.

File: lldb/source/Plugins/SymbolFile/NativePDB/SymbolFileNativePDB.cpp (hunk 2 above)
Comment (JDevlieghere) [reply lldb/source/Plugins/SymbolFile/NativePDB/SymbolFileNativePDB.cpp:1820]: Like Pavel I understood the comment as if we'd still crash. Maybe update the wording to make it explicit that that's the reason we don't handle it here.

File: lldb/source/Plugins/SymbolFile/NativePDB/SymbolFileNativePDB.h [hunk 4]
```diff
@@ -265,6 +265,7 @@ class SymbolFileNativePDB : public SymbolFileCommon {
   // UID for anonymous union and anonymous struct as they don't have entities in
   // pdb debug info.
   lldb::user_id_t anonymous_id = LLDB_INVALID_UID - 1;
+  std::optional<uint32_t> m_cu_count = 0;
```
Comment (JDevlieghere) [inline lldb/source/Plugins/SymbolFile/NativePDB/SymbolFileNativePDB.h:268]: Unrelated?

File: lldb/source/Plugins/SymbolFile/NativePDB/SymbolFileNativePDB.cpp (hunk 2 above)
Comment (ZequanWu) [reply lldb/source/Plugins/SymbolFile/NativePDB/SymbolFileNativePDB.cpp]: Updated wording.

**Review by JDevlieghere — APPROVED** [review]
LGTM. Thanks!
